project_timesheet_holidays_contract/models/hr_employee.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from datetime import datetime
+
+from odoo import _, models
+from odoo.exceptions import ValidationError
+
+
+class HrEmployee(models.Model):
+    _inherit = "hr.employee"
+
+    def _get_calendar(self, date_from=None):
+        res = super()._get_calendar(date_from)
+        if not date_from:
+            return res
+
+        contracts = (
+            self.sudo()
+            .with_context(active_test=True)
+            .env["hr.contract"]
+            .search([("employee_id", "=", self.id)])
+        )
+
+        if date_from:
+            date_from = (
+                date_from if not isinstance(date_from, datetime) else date_from.date()
+            )
+            contracts = contracts.filtered(lambda c: c.date_start <= date_from)
+
+        date_to = self.env.context.get("date_to")
+        if date_to:
+            date_to = date_to if not isinstance(date_to, datetime) else date_to.date()
+            contracts = contracts.filtered(
+                lambda c: not c.date_end or c.date_end >= date_to
+            )
+
+        if contracts:
+            # If only one contract, no matter it's state
+            if len(contracts) == 1:
+                return contracts[0].resource_calendar_id
+            # But if not, do not take cancelled ones and see what's left
+            contracts = contracts.filtered(lambda c: c.state != "cancel")
+            if contracts:
+                if len(contracts) == 1:
+                    return contracts[0].resource_calendar_id
+            # There are multiple contracts for this date period
+            raise ValidationError(_("This period overlaps multiple contracts!"))
+
+        return res
```

Comment:
If any contract is found, either a contract's calendar is returned, or an error is raised.
If no contract is found, the result of the `super()` call is returned.
But the `super()` call is executed before knowing if any contract is found, which makes that call redundant if you find any contract.
Please refactor the code so that you call `super()._get_calendar(date_from)` only if necessary, to improve performances and avoid executing code unnecessarily.

Reply:
PS: check at line 43 is useless, you can remove it.